packages/core/src/shared/crashMonitoring.ts

```diff
@@ -458,32 +462,34 @@ export class FileSystemState {
     // ------------------ Heartbeat methods ------------------
     public async sendHeartbeat() {
         const extId = this.extId
+        const filePath = this.makeStateFilePath(extId)
+        const now = this.deps.now()
+
+        let fileHandle: nodeFs.FileHandle | undefined
         try {
-            const now = this.deps.now()
             const func = async () => {
-                const filePath = this.makeStateFilePath(extId)
-
-                // We were seeing weird behavior where we possibly read an old file, even though we overwrote it.
-                // So this is a sanity check.
-                await fs.delete(filePath, { force: true })
-
-                await fs.writeFile(filePath, JSON.stringify({ ...this.ext, lastHeartbeat: now }, undefined, 4))
-
-                // Sanity check to verify the write is accessible immediately after
+                fileHandle = await nodeFs.open(filePath, 'w')
+                await fileHandle.writeFile(JSON.stringify({ ...this.ext, lastHeartbeat: now }, undefined, 4))
+                // Noticing that some file reads are not immediately available after write. `fsync` is known to address this.
+                await fileHandle!.sync()
+                await fileHandle.close()
+                fileHandle = undefined
+
+                // Sanity check to verify the latest write is accessible immediately
                 const heartbeatData = JSON.parse(await fs.readFileText(filePath)) as ExtInstanceHeartbeat
```

Comment:
Can you confirm that here, `fs` is the vscode VFS while just above we are using the nodefs. If so, that would explain why the write is not necessarily immediately seen: the vscode VFS could have a extra cache layer or other indirection.

Is it actually required/intentional to use `fs` here or should we use nodefs for this part?

Reply:
No explicit need for the VSC FS api here. I'll look to reduce its usage in case there is some layer of indirection underneath